pkg/descheduler/descheduler.go

```diff
@@ -51,11 +52,12 @@ import (
 func Run(ctx context.Context, rs *options.DeschedulerServer) error {
 	metrics.Register()
 
-	rsclient, err := client.CreateClient(rs.KubeconfigFile)
+	rsclient, eventClient, err := createClients(rs.KubeconfigFile)
```

Comment:
for my understanding, why do we need separate clients?

Reply:
The only difference between both clients is the userAgent. Which helps to separate e.g. descheduler audit logs from the events audit logs. Checking other components:
- [KCM has a single client](https://github.com/kubernetes/kubernetes/blob/a9593d634c6a053848413e600dadbf974627515f/cmd/kube-controller-manager/app/controllermanager.go#L184
)
- [kubelet has a single client](https://github.com/kubernetes/kubernetes/blob/a9593d634c6a053848413e600dadbf974627515f/cmd/kubelet/app/server.go#L450)
- [kube-proxy has both](https://github.com/kubernetes/kubernetes/blob/a9593d634c6a053848413e600dadbf974627515f/cmd/kube-proxy/app/server.go#L589)

Although in all the cases the event broadcaster is actually using only a single client for recording events:
KCM:
```go
eventBroadcaster := record.NewBroadcaster()
eventRecorder := eventBroadcaster.NewRecorder(clientgokubescheme.Scheme, v1.EventSource{Component: 
KubeControllerManagerUserAgent})
c.EventBroadcaster.StartStructuredLogging(0)
c.EventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: c.Client.CoreV1().Events("")})
defer c.EventBroadcaster.Shutdown()
```
kubelet:
```go
eventBroadcaster := record.NewBroadcaster()
kubeDeps.Recorder = eventBroadcaster.NewRecorder(legacyscheme.Scheme, v1.EventSource{Component: componentKubelet, Host: string(nodeName)})
eventBroadcaster.StartStructuredLogging(3)
eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeDeps.EventClient.Events("")})
```
kube-proxy: it creates `eventClient` but still injects `client.EventsV1()` into the event broadcaster
```go
eventClient, err := clientset.NewForConfig(kubeConfig)
eventBroadcaster := events.NewBroadcaster(&events.EventSinkImpl{Interface: client.EventsV1()})
recorder := eventBroadcaster.NewRecorder(scheme.Scheme, "kube-proxy")
...
if s.Broadcaster != nil && s.EventClient != nil {
	stopCh := make(chan struct{})
	s.Broadcaster.StartRecordingToSink(stopCh)
}
...
```

Checking other places under `k8s.io/kubernetes/staging/src/k8s.io`:
```bash
$ grep -rn "\.StartRecordingToSink" --include "*.go" --exclude "*_test.go"
sample-controller/controller.go:101:	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeclientset.CoreV1().Events("")})
client-go/tools/events/event_broadcaster.go:381:		e.eventsv1Broadcaster.StartRecordingToSink(stopCh)
client-go/tools/events/event_broadcaster.go:384:		e.coreBroadcaster.StartRecordingToSink(&typedv1core.EventSinkImpl{Interface: e.coreClient.Events("")})
legacy-cloud-providers/aws/aws.go:1423:	c.eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: c.kubeClient.CoreV1().Events("")})
legacy-cloud-providers/azure/azure.go:674:	az.eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: az.KubeClient.CoreV1().Events("")})
legacy-cloud-providers/gce/gce.go:655:	g.eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: g.client.CoreV1().Events("")})
cloud-provider/controllers/service/controller.go:114:	broadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
cloud-provider/controllers/node/node_controller.go:119:	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
cloud-provider/controllers/route/route_controller.go:109:		rc.broadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: rc.kubeClient.CoreV1().Events("")})
cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:78:	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
cloud-provider/app/controllermanager.go:148:	c.EventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: c.Client.CoreV1().Events("")})
```

Looks like kube-scheduler is the only one creating a separate event client.

Reply:
@knelasevero I will leave it up to you to make the decision whether we use a single client or separate ones. I don't have any objections here.

Reply:
Since this project is closely related to KS in heart and soul, I would slightly prefer to keep it similar to what is being done in there. But also no strong preference from my side.